### PR TITLE
Attempt to make `optimas` generators compatible with generator standard

### PR DESCRIPTION
**Pull request by RemiLehe**
Overall approach:

`optimas` generators currently call the `ask` and `tell` methods, which, under the hood, can the `_ask` and `_tell` methods.

In this PR:
- `ask`/`tell` are renamed to `ask_trials`/`tell_trials` and keep the same interface (for compatibility with the rest of the framework)
- `_ask`/`_tell` are renamed to `ask`/`tell` and adopt the new standardized interface.

This allows other packages to use the `optimas` generators. (In this case, these other packages will call the new, standardized `ask`/`tell` directly, without going through `ask_trials`/`tell_trials`.)

In order to wrap a standardized generator from another package, one would only need to make a thin wrapper around it that defines `ask_trials` and `tell_trials`.

Checks (SH):

- [x] Check [output of multitask trials](https://github.com/optimas-org/optimas/pull/249#issuecomment-2848231432)
- [x] Changed TrialParameter from `trial_index` to `ax_trial_id `
- [x] Before  `_, trial_id = self._ax_client.attach_trial(params)` could get called on an ignored trial, is this needed?
- [x] Can we "tell" all ready points or must we keep with telling one at a time (as currently does)?
 - `tell_trials` function can work on a set of points, but if do more than one at a time...
 - Saving points to file happens after the batch so does not exactly match requested frequency
 - If more than `max_evals` done, rest of batch gets given (this could be truncated with an internal check if they are not wanted).
 - More efficient to tell all ready points (small diff but consider when using a remote gen - like in APOSMM case).

### CODE REVIEW

**Comment by RemiLehe** [comment]
The failing tests are all in `test_ax_generators.py`:
Screenshot 2025-03-24 at 9 10 35 AM
Several of them seem to be related to 
```
ValueError: Could not create trials from given data because the required fields ['ax_trial_id'] are missing.
```

**Comment by shuds13** [comment]
### Optimas trial ID and Ax trial ID analysis.

**TLDR**: When the generator produces an ID “_id” (according to the standard),  this will be stored in Optimus trials as trial.gen_id, and returned to the generator when ingest is called. Optimas maintains a separate trial index (trial.index).

---

The standard says the generator can return an `_id`  in addition to the declared variables when `suggest` is called. When we give back to the gen, if `_id`  it's declared, it should match the value that came out originally.

In the Ax examples,  the generator currently returns `ax_trial_id`, which Optimus stores in its own trial (`trial.ax_trial_id`). However, after new  points have been generated,  Optimus gives them all an index,  which seems to simply enumerate from zero (`trial.index`). Before giving back to the generator this is automatically converted to `_id`. So the generator is given back both `_id` and `ax_trial_id`).

So the key question I think is whether we can expect generator ID and Optimas ID  to always match. In that case I could set the Optimus trial index to the generator ID if it exists and otherwise create the enumerated IDs.

OR

If we want to maintain separate IDs because they might differ, then I would do the following.

The ‘_id’ that comes out of the generator is stored in the Optimus trial as `trial.gen_id`. The Optimus index will be generated as before (`trial.index`). However, the to_dict function that prepares trial data to be given back to the generator will use:

 "_id": self.gen_id or self._index,

Or just 

 "_id": self.gen_id

which may be safer so as to avoid getting the two muddled - if unexpected points were given to the generator they would presumably come without IDs and, in general, if the generator doesn't produce an ID then you don't need to give one back to it.

That said, unexpected points don't seem to have an Optimus index even.

What about incorporate_history?

When you attach a previous history Optimus loads in trial IDs,  whereas the gen ID (atleast ax_trial_id) is only present for points created in this run. So this would create an offset between trial ID and gen ID.

So I think maintaining both is the way to go.

File: optimas/generators/base.py [hunk 1]
```diff
@@ -250,7 +257,11 @@ def tell(
             incorporating the evaluated trials. By default ``True``.
 
         """
-        self._tell(trials)
+        # Feed data to the generator, using the standardized API
+        self.ignore_trials(trials)
+        self.ingest([trial.to_dict() for trial in trials])
```
Comment (shuds13) [inline optimas/generators/base.py:262]: Should we only call in ingest when the trial evaluation was successful?

Is currently called even when the evaluation has failed, which results in `KeyError` In the  libEnsemble generator.
Evaluation parameters are not mapped to a trial unless the trial is successful.

Comment (shuds13) [reply optimas/generators/base.py:262]: We might also want to warn when there are fail trials

**Comment by shuds13** [comment]
Check output of multitask trials

`./print_fields.py tests_output/ax_mt_history.npy -f sim_id trial_index ax_trial_index ax_arm_name`

**Original**

2 runs:

```
[( 0,  0, 0, '0_0')
 ( 1,  1, 0, '0_1')
 ( 2,  2, 0, '0_2')
 ( 3,  3, 0, '0_3')
 ( 4,  4, 0, '0_4')
 ( 5,  5, 1, '1_0')
 ( 6,  6, 1, '1_1')
 ( 7,  7, 2, '2_0')
 ( 8,  8, 2, '2_1')
 ( 9,  9, 2, '2_2')
 (10, 10, 3, '2_2')
 (11, 11, 4, '4_0')
 (12, 12, 4, '4_1')
 (13, 13, 4, '4_2')
 (14, 14, 5, '4_0')]

[( 0,  0, 0, '0_0')
 ( 1,  1, 0, '0_1')
 ( 2,  2, 0, '0_2')
 ( 3,  3, 0, '0_3')
 ( 4,  4, 0, '0_4')
 ( 5,  5, 1, '1_0')
 ( 6,  6, 1, '1_1')
 ( 7,  7, 2, '2_0')
 ( 8,  8, 2, '2_1')
 ( 9,  9, 2, '2_2')
 (10, 10, 3, '2_0')
 (11, 11, 4, '4_0')
 (12, 12, 4, '4_1')
 (13, 13, 4, '4_2')
 (14, 14, 5, '4_2')]
```

**New**

```
[( 0,  0, 0, '0_0')
 ( 1,  1, 0, '0_1')
 ( 2,  2, 0, '0_2')
 ( 3,  3, 0, '0_3')
 ( 4,  4, 0, '0_4')
 ( 5,  5, 0, '0_4')
 ( 6,  6, 1, '1_0')
 ( 7,  7, 1, '1_1')
 ( 8,  8, 2, '2_0')
 ( 9,  9, 2, '2_1')
 (10, 10, 2, '2_2')
 (11, 11, 2, '2_2')
 (12, 12, 3, '2_2')
 (13, 13, 3, '2_2')
 (14, 14, 4, '4_0')]

[( 0,  0, 0, '0_0')
 ( 1,  1, 0, '0_1')
 ( 2,  2, 0, '0_2')
 ( 3,  3, 0, '0_3')
 ( 4,  4, 0, '0_4')
 ( 5,  5, 0, '0_4')
 ( 6,  6, 1, '1_0')
 ( 7,  7, 1, '1_1')
 ( 8,  8, 2, '2_0')
 ( 9,  9, 2, '2_1')
 (10, 10, 2, '2_2')
 (11, 11, 2, '2_2')
 (12, 12, 3, '2_0')
 (13, 13, 3, '2_0')
 (14, 14, 4, '4_0')]
```

File: optimas/generators/__init__.py [hunk 2]
```diff
@@ -32,4 +32,5 @@
     "GridSamplingGenerator",
     "LineSamplingGenerator",
     "RandomSamplingGenerator",
+    "ExternalGenerator",
```
Comment (RemiLehe) [inline optimas/generators/__init__.py:35]: Would it make sense to add a test for this object in the CI?
(This could potentially be a follow-up PR.)

File: tests/test_ax_generators.py [hunk 3]
```diff
@@ -199,6 +199,75 @@ def test_ax_single_fidelity():
     make_plots(gen)
 
 
+def test_ax_single_fidelity_resume():
```
Comment (RemiLehe) [inline tests/test_ax_generators.py:202]: It seems that this function was simply moved from its previous location in this file to this new location.

If this is indeed the case, could you move it back to its previous location?

File: optimas/generators/__init__.py (hunk 2 above)
Comment (shuds13) [reply optimas/generators/__init__.py:35]: yes

File: optimas/gen_functions.py [hunk 4]
```diff
@@ -107,8 +108,12 @@ def persistent_generator(H, persis_info, gen_specs, libE_info):
                         y = calc_in[par.name][i]
                         ev = Evaluation(parameter=par, value=y)
                         trial.complete_evaluation(ev)
-                # Register trial with unknown SEM
-                generator.tell([trial])
+                generator.tell_trials([trial])
+                trials.append(trial)
+
+            # Register trials with unknown SEM
+            # generator.tell_trials(trials)  # Give as batch
```
[the remark this answers is not in the document]
Comment (shuds13) [reply optimas/gen_functions.py:115]: I updated to batch, but it throws out a few things - like checkpointing at an exact id.

File: optimas/gen_functions.py [hunk 5]
```diff
@@ -96,6 +96,7 @@ def persistent_generator(H, persis_info, gen_specs, libE_info):
             # Check how many simulations have returned
             n = len(calc_in["sim_id"])
             # Feed the latest simulation results to the generator
+            trials = []
```
Comment (RemiLehe) [inline optimas/gen_functions.py:99]:
```suggestion
```

File: optimas/generators/ax/developer/multitask.py [hunk 6]
```diff
@@ -196,10 +196,18 @@ def __init__(
...
+        # Ax trial_index and arm toegther locate a point
+        # Multiple points (Optimas trials) can share the same Ax trial_index
+        # Standard (VOCS) does not have equiv. of trial parameters -> would be variables.
+        # If want to use _id can have a mapping inside the generator, but those
+        # points that are part of same ax trial_index will not be seen in history.
         custom_trial_parameters = [
             TrialParameter("arm_name", "ax_arm_name", dtype="U32"),
             TrialParameter("trial_type", "ax_trial_type", dtype="U32"),
-            TrialParameter("trial_index", "ax_trial_index", dtype=int),
+            # SH changed from trial_index to ax_trial_id - trial_index is usually Optimas index?
```
Comment (RemiLehe) [inline optimas/generators/ax/developer/multitask.py:209]:
```suggestion
```

File: optimas/generators/ax/developer/multitask.py [hunk 7]
```diff
@@ -196,10 +196,18 @@ def __init__(
         model_save_period: Optional[int] = 5,
         model_history_dir: Optional[str] = "model_history",
     ) -> None:
+        # SH note for standardization
```
Comment (RemiLehe) [inline optimas/generators/ax/developer/multitask.py:199]:
```suggestion
```

**Review by shuds13 — APPROVED** [review]
